codegen/smithy-aws-go-codegen/src/main/java/software/amazon/smithy/aws/go/codegen/AddAwsConfigFields.java

```diff
@@ -0,0 +1,285 @@
+/*
+ * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.smithy.aws.go.codegen;
+
+import java.util.Optional;
+import java.util.Set;
+import java.util.TreeSet;
+import java.util.function.Consumer;
+import software.amazon.smithy.codegen.core.Symbol;
+import software.amazon.smithy.codegen.core.SymbolProvider;
+import software.amazon.smithy.go.codegen.GoSettings;
+import software.amazon.smithy.go.codegen.GoWriter;
+import software.amazon.smithy.go.codegen.SymbolUtils;
+import software.amazon.smithy.go.codegen.TriConsumer;
+import software.amazon.smithy.go.codegen.integration.GoIntegration;
+import software.amazon.smithy.model.Model;
+import software.amazon.smithy.utils.SetUtils;
+import software.amazon.smithy.utils.SmithyBuilder;
+
+/**
+ * Registers additional AWS specific client configuration fields
+ */
+public class AddAwsConfigFields implements GoIntegration {
+    private static final String REGION_CONFIG_NAME = "Region";
+    private static final String CREDENTIALS_CONFIG_NAME = "Credentials";
+    private static final String ENDPOINT_RESOLVER_CONFIG_NAME = "EndpointResolver";
+    private static final String HTTP_CLIENT_CONFIG_NAME = "HTTPClient";
+    private static final String LOGGER_CONFIG_NAME = "Logger";
+    private static final String LOG_LEVEL_CONFIG_NAME = "LogLevel";
+    private static final String RETRYER_CONFIG_NAME = "Retryer";
+    private static final Set<ConfigField> UNIVERSAL_FIELDS = new TreeSet<>(SetUtils.of(
```

Comment:
I don't know how this comment got dropped from my last set, but wouldn't it generally be simpler just to put the `aws.Config` struct directly into the client configuration rather than trying to break it out?

Reply:
Had a discussion yesterday to with Jason around the design some more, and we came to the conclusion that we would like to continue down this path of generating the corresponding options on the service client. The reason being is aws.Config is really a property grab bag of parameters loaded from shared config and is not always applicable to a given service client. By generating the options on the service client we can be smarter about which options are exposed/required and which ones are not. For example if a client were to support some new authentication scheme we could generate the specific type of credential provider required.

Reply:
👍